src/segment-loader.js

```diff
@@ -717,6 +717,94 @@ export default class SegmentLoader extends videojs.EventTarget {
     };
   }
 
+  abortRequestEarly_(stats) {
+    // If the player is not paused and the current playlist is not the
+    // lowestEnabledRendition, consider the possibility of aborting the current request
+    // early for an emergency downswitch
+    // TODO: Replace timeout with a boolean indicating whether this playlist is the
+    //       lowestEnabledRendition
+    if (!this.hls_.tech_.paused() && this.xhrOptions_.timeout) {
+      const firstByteReceived = Date.now() - stats.firstByteReceived;
+
+      // Wait at least 1 second before using the calculated bandwidth from the
+      // progress event to allow the bitrate to stableize
+      if (firstByteReceived > 1000) {
+        const realBandwidth = stats.bandwidth;
+        const estimatedSize =
+          this.pendingSegment_.duration * this.playlist_.attributes.BANDWIDTH;
+        const currentTime = this.currentTime_();
+        const estimatedDelay =
+          (estimatedSize - (stats.bytesReceived * 8)) / realBandwidth;
+        const buffered = this.buffered_();
+        const bufferedEnd = buffered.length ? buffered.end(buffered.length - 1) : 0;
+        const rebufferingDelay =
+          (bufferedEnd - currentTime) / this.hls_.tech_.playbackRate();
+
+        // consider aborting early if the estimated time to finish the download
+        // is larger than the estimated time until the player runs out of forward buffer
+        if (estimatedDelay > rebufferingDelay) {
```

Comment:
Do we want to allow for some variance here? I can see us going either way, since right now we have a variance on the bandwidth when choosing the playlist. That said, it might be useful to take note of that variance here. If there is an allowed variance when selecting a rendition, then we shouldn't consider one here. If there isn't, maybe we should think about having one here.